File: src/main/java/com/atsuishio/superbwarfare/datagen/ModBlockStateProvider.java
package com.atsuishio.superbwarfare.datagen;

import com.atsuishio.superbwarfare.Mod;
import com.atsuishio.superbwarfare.init.ModBlocks;
import com.google.gson.JsonObject;
import net.minecraft.core.registries.BuiltInRegistries;
import net.minecraft.data.CachedOutput;
import net.minecraft.data.DataProvider;
import net.minecraft.data.PackOutput;
import net.minecraft.resources.ResourceLocation;
import net.minecraft.world.level.block.Block;
import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.CompletableFuture;

public class ModBlockStateProvider implements DataProvider {

    private final PackOutput output;

    private final Map<ResourceLocation, JsonObject> blockStates = new HashMap<>();
    private final Map<ResourceLocation, JsonObject> models = new HashMap<>();
    private final List<ResourceLocation> simpleItemModels = new ArrayList<>();

    public ModBlockStateProvider(PackOutput output) {
        this.output = output;
    }

    @Override
    public @NotNull CompletableFuture<?> run(@NotNull CachedOutput cache) {
        generate();

        PackOutput.PathProvider blockstatePath = output.createPathProvider(PackOutput.Target.RESOURCE_PACK, "blockstates");
        PackOutput.PathProvider modelPath = output.createPathProvider(PackOutput.Target.RESOURCE_PACK, "models/block");
        PackOutput.PathProvider itemModelPath = output.createPathProvider(PackOutput.Target.RESOURCE_PACK, "models/item");

        List<CompletableFuture<?>> futures = new ArrayList<>();

        for (var entry : blockStates.entrySet()) {
            futures.add(DataProvider.saveStable(cache, entry.getValue(), blockstatePath.json(entry.getKey())));
        }
        for (var entry : models.entrySet()) {
            futures.add(DataProvider.saveStable(cache, entry.getValue(), modelPath.json(entry.getKey())));
        }
        for (var modelId : simpleItemModels) {
            JsonObject json = new JsonObject();
            json.addProperty("parent", modelId.toString());
            futures.add(DataProvider.saveStable(cache, json, itemModelPath.json(modelId)));
        }

        return CompletableFuture.allOf(futures.toArray(CompletableFuture[]::new));
    }

    private void generate() {
        horizontalBlock(ModBlocks.BARBED_WIRE, modLoc("block/barbed_wire"));
        horizontalBlock(ModBlocks.JUMP_PAD, modLoc("block/jump_pad"));
        horizontalBlock(ModBlocks.REFORGING_TABLE, modLoc("block/reforging_table"));
        horizontalBlock(ModBlocks.CONTAINER, modLoc("block/container"));
        horizontalBlock(ModBlocks.SMALL_CONTAINER, modLoc("block/small_container"));
        horizontalBlock(ModBlocks.LUCKY_CONTAINER, modLoc("block/container"));
        horizontalBlock(ModBlocks.CHARGING_STATION, modLoc("block/charging_station"));
        horizontalBlock(ModBlocks.CREATIVE_CHARGING_STATION, modLoc("block/creative_charging_station"));
        horizontalBlock(ModBlocks.VEHICLE_DEPLOYER, cubeBottomTop("vehicle_deployer",
                modLoc("block/vehicle_deployer_side"),
                modLoc("block/vehicle_deployer_bottom"),
                modLoc("block/vehicle_deployer_top")));
        horizontalBlock(ModBlocks.VEHICLE_ASSEMBLING_TABLE, modLoc("block/vehicle_assembling_table"));

        horizontalBlock(ModBlocks.AIRCRAFT_CATAPULT, cube("aircraft_catapult",
                modLoc("block/vehicle_deployer_bottom"),
                modLoc("block/aircraft_catapult_top"),
                modLoc("block/aircraft_catapult_side"),
                modLoc("block/aircraft_catapult_side"),
                modLoc("block/aircraft_catapult_side2"),
                modLoc("block/aircraft_catapult_side2")));

        directionalBlock(ModBlocks.SUPERB_ITEM_INTERFACE, cubeBottomTop("superb_item_interface",
                modLoc("block/superb_item_interface_side"),
                modLoc("block/superb_item_interface_bottom"),
                modLoc("block/superb_item_interface_top")));

        directionalBlock(ModBlocks.CREATIVE_SUPERB_ITEM_INTERFACE, cubeBottomTop("creative_superb_item_interface",
                modLoc("block/creative_superb_item_interface_side"),
                modLoc("block/creative_superb_item_interface_bottom"),
                modLoc("block/creative_superb_item_interface_top")));

        blockWithItem(ModBlocks.GALENA_ORE);
        blockWithItem(ModBlocks.DEEPSLATE_GALENA_ORE);
        blockWithItem(ModBlocks.SCHEELITE_ORE);
        blockWithItem(ModBlocks.DEEPSLATE_SCHEELITE_ORE);
        blockWithItem(ModBlocks.LEAD_BLOCK);
        blockWithItem(ModBlocks.STEEL_BLOCK);
        blockWithItem(ModBlocks.TUNGSTEN_BLOCK);
        blockWithItem(ModBlocks.CEMENTED_CARBIDE_BLOCK);
        blockWithItem(ModBlocks.SILVER_ORE);
        blockWithItem(ModBlocks.DEEPSLATE_SILVER_ORE);
        blockWithItem(ModBlocks.SILVER_BLOCK);

        simpleBlock(ModBlocks.FUMO_25, modLoc("block/fumo_25"));
    }

    private void simpleBlock(Block block, ResourceLocation model) {
        String key = BuiltInRegistries.BLOCK.getKey(block).getPath();
        JsonObject bs = new JsonObject();
        JsonObject variants = new JsonObject();
        JsonObject variant = new JsonObject();
        variant.addProperty("model", model.toString());
        variants.add("", variant);
        bs.add("variants", variants);
        blockStates.put(Mod.loc(key), bs);
    }

    private void horizontalBlock(Block block, ResourceLocation model) {
        String key = BuiltInRegistries.BLOCK.getKey(block).getPath();
        JsonObject bs = new JsonObject();
        JsonObject variants = new JsonObject();

        addVariant(variants, "facing=north", model, 0);
        addVariant(variants, "facing=south", model, 180);
        addVariant(variants, "facing=west", model, 270);
        addVariant(variants, "facing=east", model, 90);

        bs.add("variants", variants);
        blockStates.put(Mod.loc(key), bs);
        simpleItemModel(block);
    }

    private void directionalBlock(Block block, ResourceLocation model) {
        String key = BuiltInRegistries.BLOCK.getKey(block).getPath();
        JsonObject bs = new JsonObject();
        JsonObject variants = new JsonObject();

        addVariant(variants, "facing=north", model, 0);
        addVariant(variants, "facing=east", model, 90);
        addVariant(variants, "facing=south", model, 180);
        addVariant(variants, "facing=west", model, 270);

        bs.add("variants", variants);
        blockStates.put(Mod.loc(key), bs);
        simpleItemModel(block);
    }

    private void addVariant(JsonObject variants, String key, ResourceLocation model, int y) {
        JsonObject v = new JsonObject();
        v.addProperty("model", model.toString());
        if (y != 0) v.addProperty("y", y);
        variants.add(key, v);
    }

    private void blockWithItem(Block block) {
        ResourceLocation model = cubeAll(block);
        simpleBlock(block, model);
        simpleItemModel(block);
    }

    private ResourceLocation cubeAll(Block block) {
        String name = BuiltInRegistries.BLOCK.getKey(block).getPath();
        ResourceLocation id = Mod.loc(name);
        JsonObject json = new JsonObject();
        json.addProperty("parent", "minecraft:block/cube_all");
        JsonObject textures = new JsonObject();
        textures.addProperty("all", modLoc("block/" + name).toString());
        json.add("textures", textures);
        models.put(id, json);
        return ResourceLocation.fromNamespaceAndPath(Mod.MODID, "block/" + name);
    }

    private ResourceLocation cubeBottomTop(String name, ResourceLocation side, ResourceLocation bottom, ResourceLocation top) {
        ResourceLocation id = Mod.loc(name);
        JsonObject json = new JsonObject();
        json.addProperty("parent", "minecraft:block/cube_bottom_top");
        JsonObject textures = new JsonObject();
        textures.addProperty("side", side.toString());
        textures.addProperty("bottom", bottom.toString());
        textures.addProperty("top", top.toString());
        textures.addProperty("particle", bottom.toString());
        json.add("textures", textures);
        models.put(id, json);
        return ResourceLocation.fromNamespaceAndPath(Mod.MODID, "block/" + name);
    }

    private ResourceLocation cube(String name, ResourceLocation down, ResourceLocation up, ResourceLocation north, ResourceLocation south, ResourceLocation east, ResourceLocation west) {
        ResourceLocation id = Mod.loc(name);
        JsonObject json = new JsonObject();
        json.addProperty("parent", "minecraft:block/cube");
        JsonObject textures = new JsonObject();
        textures.addProperty("down", down.toString());
        textures.addProperty("up", up.toString());
        textures.addProperty("north", north.toString());
        textures.addProperty("south", south.toString());
        textures.addProperty("east", east.toString());
        textures.addProperty("west", west.toString());
        json.add("textures", textures);
        models.put(id, json);
        return ResourceLocation.fromNamespaceAndPath(Mod.MODID, "block/" + name);
    }

    private void simpleItemModel(Block block) {
        String name = BuiltInRegistries.BLOCK.getKey(block).getPath();
        simpleItemModels.add(ResourceLocation.fromNamespaceAndPath(Mod.MODID, name));
    }

    private ResourceLocation modLoc(String path) {
        return Mod.loc(path);
    }

    @Override
    public @NotNull String getName() {
        return "Superb Warfare Block States";
    }
}

File: src/main/kotlin/com/atsuishio/superbwarfare/datagen/ModBlockStateProvider.kt
package com.atsuishio.superbwarfare.datagen

import com.atsuishio.superbwarfare.Mod
import com.atsuishio.superbwarfare.Mod.Companion.loc
import com.atsuishio.superbwarfare.init.ModBlocks
import com.google.gson.JsonObject
import net.minecraft.core.Direction
import net.minecraft.core.registries.BuiltInRegistries
import net.minecraft.data.CachedOutput
import net.minecraft.data.DataProvider
import net.minecraft.data.PackOutput
import net.minecraft.resources.ResourceLocation
import net.minecraft.util.StringRepresentable
import net.minecraft.world.level.block.Block
import net.minecraft.world.level.block.state.BlockState
import net.minecraft.world.level.block.state.properties.BlockStateProperties
import java.util.concurrent.CompletableFuture

class ModBlockStateProvider(private val output: PackOutput) : DataProvider {
    private val blockStates = linkedMapOf<ResourceLocation, JsonObject>()
    private val models = linkedMapOf<ResourceLocation, JsonObject>()

    override fun run(pOutput: CachedOutput): CompletableFuture<*> {
        blockStates.clear()
        models.clear()
        registerStatesAndModels()

        val blockStatePath = output.createPathProvider(PackOutput.Target.RESOURCE_PACK, "blockstates")
        val modelPath = output.createPathProvider(PackOutput.Target.RESOURCE_PACK, "models")
        return CompletableFuture.allOf(
            *(blockStates.map { (id, json) -> DataProvider.saveStable(pOutput, json, blockStatePath.json(id)) } +
                models.map { (id, json) -> DataProvider.saveStable(pOutput, json, modelPath.json(id)) })
                .toTypedArray()
        )
    }

    override fun getName(): String {
        return "${Mod.MODID} block states"
    }

    private fun registerStatesAndModels() {
        horizontalBlock(ModBlocks.BARBED_WIRE, blockModel("barbed_wire"))
        horizontalBlock(ModBlocks.JUMP_PAD, blockModel("jump_pad"))
        horizontalBlock(ModBlocks.REFORGING_TABLE, blockModel("reforging_table"))
        horizontalBlock(ModBlocks.CONTAINER, blockModel("container"))
        horizontalBlock(ModBlocks.SMALL_CONTAINER, blockModel("small_container"))
        horizontalBlock(ModBlocks.LUCKY_CONTAINER, blockModel("container"))
        horizontalBlock(ModBlocks.CHARGING_STATION, blockModel("charging_station"))
        horizontalBlock(ModBlocks.CREATIVE_CHARGING_STATION, blockModel("creative_charging_station"))
        horizontalBlock(
            ModBlocks.VEHICLE_DEPLOYER,
            cubeBottomTop(
                "vehicle_deployer", texture("vehicle_deployer_side"),
                texture("vehicle_deployer_bottom"), texture("vehicle_deployer_top"),
                texture("vehicle_deployer_bottom")
            )
        )
        horizontalBlock(ModBlocks.VEHICLE_ASSEMBLING_TABLE, blockModel("vehicle_assembling_table"))
        horizontalBlock(ModBlocks.BLUEPRINT_RESEARCH_TABLE, blockModel("blueprint_research_table"))
        simpleBlock(
            ModBlocks.BIOGAS_GENERATOR,
            cubeBottomTop(
                "biogas_generator",
                texture("biogas_generator_side"),
                texture("biogas_generator_bottom"),
                texture("biogas_generator_top"),
                texture("biogas_generator_side")
            )
        )

        horizontalBlock(
            ModBlocks.AIRCRAFT_CATAPULT,
            cube(
                "aircraft_catapult",
                texture("vehicle_deployer_bottom"),
                texture("aircraft_catapult_top"),
                texture("aircraft_catapult_side"),
                texture("aircraft_catapult_side"),
                texture("aircraft_catapult_side2"),
                texture("aircraft_catapult_side2"),
                texture("aircraft_catapult_top")
            )
        )

        horizontalBlock(
            ModBlocks.CATAPULT_CONTROLLER,
            cube(
                "catapult_controller",
                texture("vehicle_deployer_bottom"),
                texture("aircraft_catapult_controller_top"),
                texture("aircraft_catapult_controller_side"),
                texture("aircraft_catapult_controller_side"),
                texture("aircraft_catapult_side2"),
                texture("aircraft_catapult_side2"),
                texture("aircraft_catapult_top")
            )
        )

        directionalBlock(
            ModBlocks.SUPERB_ITEM_INTERFACE,
            cubeBottomTop(
                "superb_item_interface",
                texture("superb_item_interface_side"),
                texture("superb_item_interface_bottom"),
                texture("superb_item_interface_top"),
                texture("superb_item_interface_bottom")
            )
        )

        directionalBlock(
            ModBlocks.CREATIVE_SUPERB_ITEM_INTERFACE,
            cubeBottomTop(
                "creative_superb_item_interface",
                texture("creative_superb_item_interface_side"),
                texture("creative_superb_item_interface_bottom"),
                texture("creative_superb_item_interface_top"),
                texture("creative_superb_item_interface_bottom")
            )
        )

        blockWithItem(ModBlocks.GALENA_ORE)
        blockWithItem(ModBlocks.DEEPSLATE_GALENA_ORE)
        blockWithItem(ModBlocks.SCHEELITE_ORE)
        blockWithItem(ModBlocks.DEEPSLATE_SCHEELITE_ORE)
        blockWithItem(ModBlocks.LEAD_BLOCK)
        blockWithItem(ModBlocks.STEEL_BLOCK)
        blockWithItem(ModBlocks.TUNGSTEN_BLOCK)
        blockWithItem(ModBlocks.CEMENTED_CARBIDE_BLOCK)
        blockWithItem(ModBlocks.SILVER_ORE)
        blockWithItem(ModBlocks.DEEPSLATE_SILVER_ORE)
        blockWithItem(ModBlocks.SILVER_BLOCK)
        blockWithItem(ModBlocks.RAW_GALENA_BLOCK)
        blockWithItem(ModBlocks.RAW_SCHEELITE_BLOCK)
        blockWithItem(ModBlocks.RAW_SILVER_BLOCK)

        simpleBlock(ModBlocks.FUMO_25, blockModel("fumo_25"))
    }

    private fun simpleBlock(block: Block, model: ResourceLocation) {
        variantsBlock(block, model, RotationMode.NONE)
    }

    private fun horizontalBlock(block: Block, model: ResourceLocation) {
        variantsBlock(block, model, RotationMode.HORIZONTAL)
    }

    private fun directionalBlock(block: Block, model: ResourceLocation) {
        variantsBlock(block, model, RotationMode.DIRECTIONAL)
    }

    private fun variantsBlock(block: Block, model: ResourceLocation, rotationMode: RotationMode) {
        val variants = JsonObject()
        for (state in block.stateDefinition.possibleStates) {
            val variant = JsonObject()
            val rotation = rotation(state, rotationMode)
            variant.addProperty("model", model.toString())
            if (rotation.first != 0) {
                variant.addProperty("x", rotation.first)
            }
            if (rotation.second != 0) {
                variant.addProperty("y", rotation.second)
            }
            variants.add(variantKey(state), variant)
        }

        val json = JsonObject()
        json.add("variants", variants)
        addBlockState(block, json)
    }

    private fun blockWithItem(block: Block) {
        val path = blockPath(block)
        val model = cubeAll(path)
        simpleBlock(block, model)
        addModel(loc("item/$path"), model(model))
    }

    private fun cubeAll(name: String): ResourceLocation {
        return addModel(
            loc("block/$name"),
            model(ResourceLocation.withDefaultNamespace("block/cube_all"), mapOf("all" to texture(name)))
        )
    }

    private fun cubeBottomTop(
        name: String,
        side: ResourceLocation,
        bottom: ResourceLocation,
        top: ResourceLocation,
        particle: ResourceLocation
    ): ResourceLocation {
        return addModel(
            loc("block/$name"),
            model(
                ResourceLocation.withDefaultNamespace("block/cube_bottom_top"),
                linkedMapOf("bottom" to bottom, "particle" to particle, "side" to side, "top" to top)
            )
        )
    }

    private fun cube(
        name: String,
        down: ResourceLocation,
        up: ResourceLocation,
        north: ResourceLocation,
        south: ResourceLocation,
        east: ResourceLocation,
        west: ResourceLocation,
        particle: ResourceLocation
    ): ResourceLocation {
        return addModel(
            loc("block/$name"),
            model(
                ResourceLocation.withDefaultNamespace("block/cube"),
                linkedMapOf(
                    "down" to down,
                    "east" to east,
                    "north" to north,
                    "particle" to particle,
                    "south" to south,
                    "up" to up,
                    "west" to west
                )
            )
        )
    }

    private fun model(parent: ResourceLocation, textures: Map<String, ResourceLocation> = emptyMap()): JsonObject {
        val json = JsonObject()
        json.addProperty("parent", parent.toString())
        if (textures.isNotEmpty()) {
            val textureJson = JsonObject()
            textures.forEach { (key, value) -> textureJson.addProperty(key, value.toString()) }
            json.add("textures", textureJson)
        }
        return json
    }

    private fun addBlockState(block: Block, json: JsonObject) {
        val id = blockKey(block)
        check(blockStates.put(id, json) == null) { "Duplicate blockstate: $id" }
    }

    private fun addModel(id: ResourceLocation, json: JsonObject): ResourceLocation {
        check(models.put(id, json) == null) { "Duplicate model: $id" }
        return id
    }

    private fun rotation(state: BlockState, mode: RotationMode): Pair<Int, Int> {
        return when (mode) {
            RotationMode.NONE -> 0 to 0
            RotationMode.HORIZONTAL -> 0 to horizontalY(state.getValue(BlockStateProperties.HORIZONTAL_FACING))
            RotationMode.DIRECTIONAL -> directionalRotation(state.getValue(BlockStateProperties.FACING))
        }
    }

    private fun horizontalY(direction: Direction): Int {
        return when (direction) {
            Direction.EAST -> 90
            Direction.SOUTH -> 180
            Direction.WEST -> 270
            else -> 0
        }
    }

    private fun directionalRotation(direction: Direction): Pair<Int, Int> {
        return when (direction) {
            Direction.DOWN -> 180 to 0
            Direction.UP -> 0 to 0
            Direction.NORTH -> 90 to 0
            Direction.EAST -> 90 to 90
            Direction.SOUTH -> 90 to 180
            Direction.WEST -> 90 to 270
        }
    }

    private fun variantKey(state: BlockState): String {
        return state.values.entries.joinToString(",") { (property, value) ->
            "${property.name}=${valueName(value)}"
        }
    }

    private fun valueName(value: Comparable<*>): String {
        return if (value is StringRepresentable) value.serializedName else value.toString()
    }

    private fun blockModel(name: String): ResourceLocation {
        return loc("block/$name")
    }

    private fun texture(name: String): ResourceLocation {
        return loc("block/$name")
    }

    private fun blockKey(block: Block): ResourceLocation {
        return BuiltInRegistries.BLOCK.getKey(block)
    }

    private fun blockPath(block: Block): String {
        return blockKey(block).path
    }

    private enum class RotationMode {
        NONE,
        HORIZONTAL,
        DIRECTIONAL
    }
}
